Add pairs plot for arbitrary quantities

vpratz opened the pull request:
This PR adds a `pairs_quantity` plotting function that provides a pairs plot of an arbitrary quantity.

For a trained `BasicWorkflow`, one could use it like that:

```python
def identity(x, *args, **kwargs):
    return x

test_data = simulator.sample(500)
samples = workflow.sample(num_samples=10, conditions=test_data)
values = bf.diagnostics.metrics.posterior_contraction(samples, test_data, aggregation=identity)
fig = bf.diagnostics.pairs_quantity(values["values"], test_data)
```

The current implementation only features structures as returned by the `values` field of `posterior_contraction`, but should be generalizable to other settings. A similar non-paired plot (containing only the diagonal) could be nice as well.

Below is an example of the plot:

<img width="540" height="542" alt="image" src="https://github.com/user-attachments/assets/94cd19c7-9b18-47e0-abae-bc335f341759" />

paul-buerkner commented:
Thank you! That looks good to be as a basic implementation. The documentation doesn't mention the required dimensionality of values. Could this be clarified?

elseml commented:
Nice! I'd add a "Contraction" label to the colorbar and the y-axis of the diagonal for consistency with other plots and also to make the equivalence between the colorbar and the right-hand side y-axes clear. The cleanest solution would imo be to place them at the colorbar + at the right side of only the last diagonal element's y-axis, since it prevents cluttering the main plot.

paul-buerkner commented:
I have thought about the same of whether we should allow functions too. If we aim to also support "test_quantities" similar to the SBC diagnostics, we would have to provide "estimates" anyway. If `quantity` (or however called) is an an array or dict of array, ignore estimates and test_quantites. If `quantity` is a function then require estimates and support test_quantities. We can of course support test_quantities later, since it is not high priority. However, we should think about it already now because "quantity" and "test_quantities" will have clashing terminology.

vpratz commented:
@paul-buerkner Thanks a lot for the input. The many different options make it a bit convoluted, but I have now implemented the following:
- `values` as array
- `values` as dict from a metric function with `aggregation=None`
- `values` as function that has to return one of the above

If have also included the `test_quantities`, here as well we have two cases:
1. the quantity is scalar, so we do not require `estimates` as we do not have to calculate it per quantity, they are only different axes
2. the quantity is a vector with one entry per variable, we require estimates and a function to calculate the values

Here we could skip 1. to simplify the code, this would then correspond to the description in your comment, but I think having the more general implementation does not hurt. I have created helper functions to avoid code duplication, I'm not sure yet where we want to put them.

Let me know what you think.

<details>
<summary>Example</summary>

```python
fig = bf.diagnostics.pairs_quantity(
    values=bf.diagnostics.posterior_contraction,
    targets=test_data,
    estimates=samples,
    test_quantities=test_quantities,
    variable_names=[r"$\theta_1$", r"$\theta_2$"],
)
```

<img width="1054" height="1063" alt="image" src="https://github.com/user-attachments/assets/0af1ff43-2da1-4b96-8d45-445a67c7eec7" />
</details>

vpratz commented:
Thanks for taking a look. Regarding your questions:

- Yes, aggregation=None should behave like an identity, i.e., not aggregate over datasets but instead return the individual values, as we require them for plotting. If I read it correctly, for now this only makes sense for `posterior_contraction`. It would be viable for example for z scores as well, but as far as I can tell we have not implemented that as a metric yet.
- I just discussed with @han-ol, and it probably makes sense to remove option 1. The idea was basically, if we have a value that is a property of the dataset, e.g. its log-prob, we do not need to recalculate it even if `test_quantities` is present, which might be nice if the quantity is costly or cannot easily be recalculated. `test_quantities` would then just supply new axes to plot against. This can become muddy quite quickly though (imagine the user wants to calculate the mean posterior contraction over all variables, which would be a scalar value that should change), and it is an edge case that probably no one will ever use, so for the sake of maintainability and clarity I would remove it. Do you agree?

paul-buerkner commented:
- Okay. I think we should add a z-score metric too but of course this will be a different issue.
- Agreed.

vpratz commented:
Thanks a lot for your input. I think the functions are now ready for a final review, @paul-buerkner.
I added a few tests and updated a few details to incorporate the changes discussed above.
